docs/cardano-monads.rst

```diff
@@ -0,0 +1,894 @@
+requires two natural transformations: ``n ~> m`` and ``m ~> n``.
+
+First, providing these natural transformations is an inconvenience (the ``n ~> m`` one
+tends to be a mere lift, but the second one must do certain tricks to unlift).
+Second, it introduces a potential for horrible, subtle bugs. As was said, right
+now we misuse this helper, in particular because we call it in a runner,
+``runRealMode``; therefore, the unlift natural transformation must reconstruct the
+monadic context from what it has at transformer stack initialization site, not
+at conversation fork site; therefore, any local modifications to the monad
+transformer stack are *not* reflected in forked conversations (such as ``ReaderT``
+local).
+
+To put it into concrete terms, here is a code demonstration::
+
+    x <- ask
+    local f $ enqueueMsg msg $ \_ _ -> (:|[]) $ do
+      y <- ask
+
+What do you think, is ``y`` equal to ``x`` or ``f x``? The way we currently use
+``hoistSendActions``, it will be equal to ``x`` (completely oblivious to ``f``). This
+might be not at all what a programmer would expect.
+
+The moral of this story is that, perhaps, explicit dictionaries are a bad design
+because it's very easy to misuse them. A good effect system should take care of
+things like this. (But perhaps it's an over-genralization and it's only bad to unlift,
+whereas lifting is straightforward).
+
+Usage of Effect Context Sums
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Rather than using a method dictionary, in order to avoid lifting/unlifting
+altogether, we use effect context sums. For example, consider the following
+hypothetical method dictionary::
+
+    data MonadDbGet m = MonadDbGet
+      { dbGet :: DBTag -> ByteString -> m (Maybe ByteString)
+      }
+
+There are two implementations for it::
+
+    dbGetRocks :: MonadRealDB ctx m => DBTag -> ByteString -> m (Maybe ByteString)
+    dbGetRocks tag key = getDBByTag tag >>= rocksGetBytes key
+
+    dbGetPure :: MonadPureDB ctx m => DBTag -> ByteString -> m (Maybe ByteString)
+    dbGetPure (tagToLens -> l) key =
+        view (l . at key) <$> (view (lensOf @DBPureVar) >>= readIORef)
+
+In case of ``dbGetRocks``, the necessary context for this operation is having
+``NodeDBs`` (implied by ``MonadRealDB``). In case of ``dbGetPure``, the
+necessary context is having ``DBPureVar``.
+
+Rather than put ``MonadDbGet`` effect dictionary into the ``ReaderT`` environment, we
+construct a sum of all possible context values::
+
+    data DBSum = RealDB NodeDBs | PureDB DBPureVar
+
+This sum is used to implement a method that can use either one of the
+implementations::
+
+    eitherDB
+        :: (MonadReader ctx m, HasLens DBSum ctx DBSum)
+        => ReaderT NodeDBs m a -> ReaderT DBPureVar m a -> m a
+    eitherDB ract pact = view (lensOf @DBSum) >>= \case
+        RealDB dbs -> runReaderT ract dbs
+        PureDB pdb -> runReaderT pact pdb
+
+    dbGetSum
+        :: MonadDBSum ctx m
```

Comment:
isn't it better also to define `MonadDBSum` above?

Reply:
I can't copy all code, the example must be short. It's just some context needed for implementation. I thought about defining it, but look, it's big

```
type MonadDBSum ctx m =
    ( MonadReader ctx m
    , HasLens DBSum ctx DBSum
    , MonadMask m
    , MonadBaseControl IO m
    , MonadIO m
    , HasConfiguration
    )
```